server/images.go

```diff
@@ -300,250 +263,197 @@ func CreateModel(ctx context.Context, name, modelFileDir, quantization string, c
-				config.SetModelType(format.HumanNumber(ggml.KV().ParameterCount()))
-				config.SetFileType(ggml.KV().FileType())
+					if err := llm.Quantize(blob, temp.Name(), ftype); err != nil {
+						err2 = err
+						return false
+					}
 
-				mediatype := mediatype
-				if ggml.KV().Architecture() == "clip" {
-					mediatype = "application/vnd.ollama.image.projector"
+					layer, err = NewLayer(temp, layer.MediaType)
+					if err != nil {
+						err2 = err
+						return false
+					}
 				}
 
-				sr := io.NewSectionReader(bin, offset, size)
-				layer, err := NewLayer(sr, mediatype)
-				if err != nil {
-					return err
+				if ggml != nil {
+					config.ModelFormat = cmp.Or(config.ModelFormat, ggml.Name())
+					config.ModelFamily = cmp.Or(config.ModelFamily, ggml.KV().Architecture())
+					config.ModelType = cmp.Or(config.ModelType, format.HumanNumber(ggml.KV().ParameterCount()))
+					config.FileType = cmp.Or(config.FileType, ggml.KV().FileType())
+					config.ModelFamilies = append(config.ModelFamilies, ggml.KV().Architecture())
 				}
 
-				layers.Add(layer)
+				layers = append(layers, layer)
+				return true
+			})
 
-				offset += size
+			for _, tempfile := range tempfiles {
+				defer tempfile.Close()
+				defer os.Remove(tempfile.Name())
 			}
-		case "adapter":
-			if strings.HasPrefix(c.Args, "@") {
-				blobPath, err := GetBlobsPath(strings.TrimPrefix(c.Args, "@"))
-				if err != nil {
-					return err
-				}
 
-				c.Args = blobPath
+			if err2 != nil {
+				return err2
 			}
-
-			fn(api.ProgressResponse{Status: "creating adapter layer"})
-			bin, err := os.Open(realpath(modelFileDir, c.Args))
+		case "license", "template", "system":
+			blob := strings.NewReader(c.Args)
+			layer, err := NewLayer(blob, mediatype)
 			if err != nil {
 				return err
 			}
-			defer bin.Close()
 
-			_, size, err := llm.DecodeGGML(bin)
-			if err != nil {
-				return err
+			if c.Name != "license" {
+				// replace
+				layers = slices.DeleteFunc(layers, func(layer *Layer) bool {
+					return layer.MediaType == mediatype
+				})
 			}
 
-			sr := io.NewSectionReader(bin, 0, size)
-			layer, err := NewLayer(sr, mediatype)
+			layers = append(layers, layer)
+		case "message":
+			role, content, ok := strings.Cut(c.Args, ": ")
+			if !ok {
+				return fmt.Errorf("invalid message: %s", c.Args)
+			}
+
+			messages = append(messages, &api.Message{Role: role, Content: content})
+		default:
+			ps, err := api.FormatParams(map[string][]string{c.Name: {c.Args}})
 			if err != nil {
 				return err
 			}
 
-			layers.Add(layer)
-		case "license":
-			fn(api.ProgressResponse{Status: "creating license layer"})
+			for k, v := range ps {
+				if ks, ok := parameters[k].([]string); ok {
+					parameters[k] = append(ks, v.([]string)...)
+				} else if vs, ok := v.([]string); ok {
+					parameters[k] = vs
+				} else {
+					parameters[k] = v
+				}
+			}
+		}
+	}
+
+	var err2 error
+	layers = slices.DeleteFunc(layers, func(layer *Layer) bool {
```

Comment:
the anonymous function really gets buried in here when you're trying to read the code. Can you make this a named function? `pruneLayers()` or something along those lines? `CreateModel()` is already huge by this point.

Reply:
this loses a lot of context if its split into a top level function since deleting the layer is specific to the criteria of creating a model